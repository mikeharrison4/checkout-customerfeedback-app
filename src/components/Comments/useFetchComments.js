import { useQuery } from 'react-query';
import axios from 'axios';

const fetchComments = async () => await axios.get('http://localhost:3001/comments');

export const useFetchComments = (setComments) => {
  const {
    isLoading,
    isError,
  } = useQuery('comments', fetchComments, {
    retry: 1,
    onSuccess: (comments) => setComments([...comments.data.reverse()]),
  });

  return { isLoading, isError };
};