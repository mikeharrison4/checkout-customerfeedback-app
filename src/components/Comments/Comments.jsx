import React, { useState } from 'react';
import axios from 'axios';
import RatingTrends from './RatingTrends/RatingTrends';
import LatestComments from './LatestComments/LatestComments';
import LoadingSpinner from '../Util/LoadingSpinner';
import content from '../../data/comments.json';
import Error from '../Util/Error';
import { useFetchComments } from './useFetchComments';

const generateRatings = (comments) => comments.reduce((prev, current) => ({
  ...prev,
  [current.rating]: (prev[current.rating] || 0) + 1
}), []);

const Comments = () => {
  const [comments, setComments] = React.useState([]);

  const { isLoading, isError } = useFetchComments(setComments);

  if (isLoading) {
    return (
      <div className="flex justify-center items-center">
        <LoadingSpinner />
      </div>
    );
  }

  if (isError) {
    return (
      <div className="flex justify-center items-center">
        <Error errorMessage={content.loadError} />
      </div>
    );
  }

  return (
    <>
      <RatingTrends commentRatings={generateRatings(comments)} />
      <LatestComments comments={comments} />
    </>
  );
};

export default Comments;
