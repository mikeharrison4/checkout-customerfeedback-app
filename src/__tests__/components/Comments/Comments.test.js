import React from 'react';
import { shallow } from 'enzyme';
import Comments from '../../../components/Comments/Comments';
import { useFetchComments } from '../../../components/Comments/useFetchComments';

jest.mock('../../../components/Comments/useFetchComments', () => ({
  useFetchComments: jest.fn(),
}));

let useState;

describe('Comments.jsx', () => {
  const setState = jest.fn();
  const comments = [{
    id: 1,
    name: 'mockName',
    email: 'mockEmail@mockaroo.com',
    rating: 5,
    comment: 'mockComment'
  }];

  beforeEach(() => {
    useFetchComments.mockImplementation(() => ({
      isLoading: false,
      isError: false,
    }));
    useState = jest.spyOn(React, 'useState');
    useState.mockImplementation(() => [comments, setState]);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  
  it('should render correctly when isLoading and isError are both false', () => {
    const renderedComponent = shallow(<Comments />);
    expect(useFetchComments).toHaveBeenCalledTimes(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render correctly when isLoading is true', () => {
    useFetchComments.mockImplementation(() => ({
      isLoading: true
    }));
    const renderedComponent = shallow(<Comments />);
    expect(useFetchComments).toHaveBeenCalledTimes(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render correctly when isLoading is true', () => {
    useFetchComments.mockImplementation(() => ({
      isError: true
    }));
    const renderedComponent = shallow(<Comments />);
    expect(useFetchComments).toHaveBeenCalledTimes(1);
    expect(renderedComponent).toMatchSnapshot();
  });
});