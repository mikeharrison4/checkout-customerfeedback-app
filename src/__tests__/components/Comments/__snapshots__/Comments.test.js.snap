// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments.jsx should render correctly when isLoading and isError are both false 1`] = `
<Fragment>
  <RatingTrends
    commentRatings={
      Object {
        "5": 1,
      }
    }
  />
  <LatestComments
    comments={
      Array [
        Object {
          "comment": "mockComment",
          "email": "mockEmail@mockaroo.com",
          "id": 1,
          "name": "mockName",
          "rating": 5,
        },
      ]
    }
  />
</Fragment>
`;

exports[`Comments.jsx should render correctly when isLoading is true 1`] = `
<div
  className="flex justify-center items-center"
>
  <LoadingSpinner />
</div>
`;

exports[`Comments.jsx should render correctly when isLoading is true 2`] = `
<div
  className="flex justify-center items-center"
>
  <Error
    errorMessage="Comments cannot be loaded at this time, we apologies for any inconvenience."
  />
</div>
`;
